Extend Vars support.

In order to allow all Fix expression parameters to include "compile-time" [variables](https://github.com/metafacture/metafacture-core/wiki/Metamorph-User-Guide#parameters-to-metamorph-definitions), I'd like to propose the following changes:

- Introduce new (custom) Fix functions (names up for discussion):
  - `put_var("varName", "value")`: Define single variable (with variable substitution).
  - `put_vars("varName": "value", ...)`: Define multiple variables (without variable substitution¹).
- Update existing expression usages:
  - `do`²/`if`/`elsif`/`unless`: Resolve variable substitutions in parameters.

¹ No support for variable substitution in options required yet.
² Currently there's no bind that accepts any parameters.

Tried with:

https://gitlab.com/oersi/oersi-etl/-/tree/FunctionalReview/data/experimental/varTest

It does not seem to work:

```
$ ./gradlew --refresh-dependencies run --args '/home/tobias/git/oersi-etl/data/experimental/varTest/var.flux'

> Task :run
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/tobias/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-log4j12/1.7.21/7238b064d1aba20da2ac03217d700d91e02460fa/slf4j-log4j12-1.7.21.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/tobias/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.21/be4b3c560a37e69b6c58278116740db28832232c/slf4j-simple-1.7.21.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
2021-12-10 12:56:38 [INFO] [ETL:127] Running /home/tobias/git/oersi-etl/data/experimental/varTest/var.flux with 5 arguments (configure output details in log4j.properties)
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$2 (file:/home/tobias/.gradle/caches/modules-2/files-2.1/com.google.inject/guice/3.0/9d84f15fe35e2c716a02979fb62f50a29f38aefa/guice-3.0.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$2
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
2021-12-10 12:56:39 [INFO] [ETL:158] Import channel var.flux, SUCCESS: n/a, FAIL-PROCESS: n/a, FAIL-VALIDATION: n/a, FAIL-WRITE: n/a, DURATION: 00:00:01
2021-12-10 12:56:39 [ERROR] [ETL:60] 
java.lang.UnsupportedOperationException
        at java.base/java.util.Collections$UnmodifiableMap.put(Collections.java:1457)
        at org.metafacture.metafix.Metafix.putVar(Metafix.java:251)
        at org.metafacture.metafix.FixMethod$1.apply(FixMethod.java:35)
        at org.metafacture.metafix.RecordTransformer.processFunction(RecordTransformer.java:188)
        at org.metafacture.metafix.RecordTransformer.processSubexpressions(RecordTransformer.java:92)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:63)
        at org.metafacture.metafix.Metafix.endRecord(Metafix.java:139)
        at org.metafacture.json.JsonDecoder.decode(JsonDecoder.java:194)
        at org.metafacture.json.JsonDecoder.processRecord(JsonDecoder.java:147)
        at org.metafacture.json.JsonDecoder.process(JsonDecoder.java:136)
        at org.metafacture.json.JsonDecoder.process(JsonDecoder.java:44)
        at org.metafacture.io.RecordReader.emitRecord(RecordReader.java:111)
        at org.metafacture.io.RecordReader.process(RecordReader.java:100)
        at org.metafacture.io.RecordReader.process(RecordReader.java:39)
        at org.metafacture.io.FileOpener.process(FileOpener.java:100)
        at org.metafacture.io.FileOpener.process(FileOpener.java:40)
        at org.metafacture.flux.parser.StringSender.process(StringSender.java:38)
        at org.metafacture.flux.parser.Flow.start(Flow.java:110)
        at org.metafacture.flux.parser.FluxProgramm.start(FluxProgramm.java:156)
        at org.metafacture.runner.Flux.main(Flux.java:79)
        at oersi.ETL.run(ETL.java:102)
        at oersi.ETL.main(ETL.java:56)

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 29s
3 actionable tasks: 1 executed, 2 up-to-date

```

--- a/metafix/src/main/java/org/metafacture/metafix/FixMethod.java
+++ b/metafix/src/main/java/org/metafacture/metafix/FixMethod.java
@@ -30,6 +30,16 @@ enum FixMethod {
 
     // SCRIPT-LEVEL METHODS:
 
+    put_var {
+        public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
+            metafix.putVar(params.get(0), params.get(1));
+        }
+    },
+    put_vars {
+        public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
+            options.forEach(metafix::putVar);
+        }
+    },
     put_map {
         public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
             metafix.putMap(params.get(0), options);

--- a/metafix/src/main/java/org/metafacture/metafix/FixMethod.java
+++ b/metafix/src/main/java/org/metafacture/metafix/FixMethod.java
@@ -32,12 +32,12 @@ enum FixMethod {
 
     put_var {
         public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
-            metafix.putVar(params.get(0), params.get(1));
+            metafix.getVars().put(params.get(0), params.get(1));
         }
     },
     put_vars {
         public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
-            options.forEach(metafix::putVar);
+            metafix.getVars().putAll(options);
         }
     },
     put_map {

--- a/metafix/src/main/java/org/metafacture/metafix/Metafix.java
+++ b/metafix/src/main/java/org/metafacture/metafix/Metafix.java
@@ -73,14 +73,25 @@ public class Metafix implements StreamPipe<StreamReceiver>, Maps { // checkstyle
     // TODO: Use SimpleRegexTrie / WildcardTrie for wildcard, alternation and character class support
     private Record currentRecord = new Record();
     private Fix fix;
-    private Map<String, String> vars = NO_VARS;
+    private Map<String, String> vars = new HashMap<>();
     private int entityCount;
     private StreamReceiver outputStreamReceiver;
     private String recordIdentifier;
     private List<Value.Hash> entities = new ArrayList<>();
 
     public Metafix() {
-        init();
+        flattener.setReceiver(new DefaultStreamReceiver() {
+            @Override
+            public void literal(final String name, final String value) {
+                // TODO: keep flattener as option?
+                // add(currentRecord, name, value);
+            }
+        });
+    }
+
+    public Metafix(final Map<String, String> newVars) {
+        this();
+        vars.putAll(newVars);
     }
 
     public Metafix(final String fixDef) throws FileNotFoundException {
@@ -91,28 +102,17 @@ public class Metafix implements StreamPipe<StreamReceiver>, Maps { // checkstyle
         this(fixDef.endsWith(".fix") ? new FileReader(fixDef) : new StringReader(fixDef), vars);
     }
 
-    public Metafix(final Reader morphDef) {
-        this(morphDef, NO_VARS);
+    public Metafix(final Reader fixDef) {
+        this(fixDef, NO_VARS);
     }
 
     public Metafix(final Reader fixDef, final Map<String, String> vars) {
+        this(vars);
         fix = FixStandaloneSetup.parseFix(fixDef);
-        this.vars = vars;
-        init();
     }
 
     public List<Expression> getExpressions() {
         return expressions;
-    }
-
-    private void init() {
-        flattener.setReceiver(new DefaultStreamReceiver() {
-            @Override
-            public void literal(final String name, final String value) {
-                // TODO: keep flattener as option?
-                // add(currentRecord, name, value);
-            }
-        });
     }
 
     @Override
@@ -241,16 +241,6 @@ public class Metafix implements StreamPipe<StreamReceiver>, Maps { // checkstyle
 
     public Map<String, String> getVars() {
         return vars;
-    }
-
-    public void putVar(final String key, final String value) {
-        try {
-            vars.put(key, value);
-        }
-        catch (final UnsupportedOperationException e) {
-            vars = new HashMap<>(vars);
-            vars.put(key, value);
-        }
     }
 
     public Record getCurrentRecord() {

--- a/metafix/src/main/java/org/metafacture/metafix/Metafix.java
+++ b/metafix/src/main/java/org/metafacture/metafix/Metafix.java
@@ -244,11 +244,13 @@ public class Metafix implements StreamPipe<StreamReceiver>, Maps { // checkstyle
     }
 
     public void putVar(final String key, final String value) {
-        if (vars == NO_VARS) {
-            vars = new HashMap<>();
+        try {
+            vars.put(key, value);
         }
-
-        vars.put(key, value);
+        catch (final UnsupportedOperationException e) {
+            vars = new HashMap<>(vars);
+            vars.put(key, value);
+        }
     }
 
     public Record getCurrentRecord() {

--- a/metafix/src/main/java/org/metafacture/metafix/Metafix.java
+++ b/metafix/src/main/java/org/metafacture/metafix/Metafix.java
@@ -96,7 +96,8 @@ public class Metafix implements StreamPipe<StreamReceiver>, Maps { // checkstyle
     }
 
     public Metafix(final Reader fixDef, final Map<String, String> vars) {
-        buildPipeline(fixDef, vars);
+        fix = FixStandaloneSetup.parseFix(fixDef);
+        this.vars = vars;
         init();
     }
 
@@ -112,11 +113,6 @@ public class Metafix implements StreamPipe<StreamReceiver>, Maps { // checkstyle
                 // add(currentRecord, name, value);
             }
         });
-    }
-
-    private void buildPipeline(final Reader fixDef, final Map<String, String> theVars) {
-        this.fix = FixStandaloneSetup.parseFix(fixDef);
-        this.vars = theVars;
     }
 
     @Override
@@ -245,6 +241,14 @@ public class Metafix implements StreamPipe<StreamReceiver>, Maps { // checkstyle
 
     public Map<String, String> getVars() {
         return vars;
+    }
+
+    public void putVar(final String key, final String value) {
+        if (vars == NO_VARS) {
+            vars = new HashMap<>();
+        }
+
+        vars.put(key, value);
     }
 
     public Record getCurrentRecord() {

--- a/metafix/src/main/java/org/metafacture/metafix/RecordTransformer.java
+++ b/metafix/src/main/java/org/metafacture/metafix/RecordTransformer.java
@@ -29,7 +29,6 @@ import org.metafacture.metafix.fix.MethodCall;
 import org.metafacture.metafix.fix.Options;
 import org.metafacture.metafix.fix.Unless;
 
-import org.eclipse.emf.common.util.EList;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -73,9 +72,13 @@ class RecordTransformer {
         return string == null ? null : StringUtil.format(string, Metafix.VAR_START, Metafix.VAR_END, false, metafix.getVars());
     }
 
+    private List<String> resolveParams(final List<String> params) {
+        return params.stream().map(this::resolveVars).collect(Collectors.toList());
+    }
+
     private void processSubexpressions(final List<Expression> expressions) {
         for (final Expression sub : expressions) {
-            final EList<String> params = sub.getParams();
+            final List<String> params = resolveParams(sub.getParams());
             if (sub instanceof Do) {
                 processBind((Do) sub, params);
             }
@@ -91,7 +94,7 @@ class RecordTransformer {
         }
     }
 
-    private void processBind(final Do theDo, final EList<String> params) {
+    private void processBind(final Do theDo, final List<String> params) {
         if (theDo.getName().equals("list")) { // TODO impl multiple binds via FixBind enum
             final Map<String, String> options = options(theDo.getOptions());
             record.findList(options.get("path"), a -> {
@@ -131,13 +134,13 @@ class RecordTransformer {
         }
     }
 
-    private void processIf(final If ifExp, final EList<String> parameters) {
+    private void processIf(final If ifExp, final List<String> params) {
         final ElsIf elsIfExp = ifExp.getElseIf();
         final Else elseExp = ifExp.getElse();
-        if (testConditional(ifExp.getName(), parameters)) {
+        if (testConditional(ifExp.getName(), params)) {
             processSubexpressions(ifExp.getElements());
         }
-        else if (elsIfExp != null && testConditional(elsIfExp.getName(), elsIfExp.getParams())) {
+        else if (elsIfExp != null && testConditional(elsIfExp.getName(), resolveParams(elsIfExp.getParams()))) {
             processSubexpressions(elsIfExp.getElements());
         }
         else if (elseExp != null) {
@@ -145,13 +148,13 @@ class RecordTransformer {
         }
     }
 
-    private void processUnless(final Unless unless, final EList<String> parameters) {
-        if (!testConditional(unless.getName(), parameters)) {
+    private void processUnless(final Unless unless, final List<String> params) {
+        if (!testConditional(unless.getName(), params)) {
             processSubexpressions(unless.getElements());
         }
     }
 
-    private boolean testConditional(final String conditional, final EList<String> params) {
+    private boolean testConditional(final String conditional, final List<String> params) {
         LOG.debug("<IF>: {} parameters: {}", conditional, params);
         boolean result = false;
         if ("exists".equals(conditional)) {
@@ -174,16 +177,15 @@ class RecordTransformer {
         // functionFactory.registerClass("not_equals", NotEquals.class);
         // functionFactory.registerClass("replace_all", Replace.class);
         // final Function function = functionFactory.newInstance(conditional,
-        // resolvedAttributeMap(parameters, theIf.getOptions()));
+        // resolvedAttributeMap(params, theIf.getOptions()));
         return result;
     }
 
     private void processFunction(final Expression expression, final List<String> params) {
         try {
             final FixMethod method = FixMethod.valueOf(expression.getName());
-            final List<String> resolvedParams = params.stream().map(this::resolveVars).collect(Collectors.toList());
             final Map<String, String> options = options(((MethodCall) expression).getOptions());
-            method.apply(metafix, record, resolvedParams, options);
+            method.apply(metafix, record, params, options);
         }
         catch (final IllegalArgumentException e) {
             throw new MetafactureException(e);

--- a/metafix/src/test/java/org/metafacture/metafix/MetafixIfTest.java
+++ b/metafix/src/test/java/org/metafacture/metafix/MetafixIfTest.java
@@ -18,6 +18,7 @@ package org.metafacture.metafix;
 
 import org.metafacture.framework.StreamReceiver;
 
+import com.google.common.collect.ImmutableMap;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.api.extension.RegisterExtension;
@@ -515,4 +516,123 @@ public class MetafixIfTest {
                 o.get().endRecord();
             });
     }
+
+    @Test
+    public void shouldResolveVariablesInIf() {
+        MetafixTestHelpers.assertFix(streamReceiver, Arrays.asList(
+                "if any_contain('name', 'Uni$[var]sity')",
+                "  add_field('type', 'Organization')",
+                "end"
+            ),
+            ImmutableMap.of("var", "ver"),
+            i -> {
+                i.startRecord("1");
+                i.literal("name", "Mary");
+                i.literal("name", "A University");
+                i.endRecord();
+
+                i.startRecord("2");
+                i.literal("name", "Mary");
+                i.literal("name", "Max");
+                i.endRecord();
+
+                i.startRecord("3");
+                i.endRecord();
+            },
+            o -> {
+                o.get().startRecord("1");
+                o.get().startEntity("name");
+                o.get().literal("1", "Mary");
+                o.get().literal("2", "A University");
+                o.get().endEntity();
+                o.get().literal("type", "Organization");
+                o.get().endRecord();
+
+                o.get().startRecord("2");
+                o.get().startEntity("name");
+                o.get().literal("1", "Mary");
+                o.get().literal("2", "Max");
+                o.get().endEntity();
+                o.get().endRecord();
+
+                o.get().startRecord("3");
+                o.get().endRecord();
+            }
+        );
+    }
+
+    @Test
+    public void shouldResolveVariablesInElsIf() {
+        MetafixTestHelpers.assertFix(streamReceiver, Arrays.asList(
+                "if any_match('name', '.*University.*')",
+                "  add_field('type', 'Organization')",
+                "elsif any_match('$[var]e', '[^ ]* [^ ]*')",
+                "  add_field('type', 'Person')",
+                "else",
+                "  add_field('type', 'Unknown')",
+                "end"
+            ),
+            ImmutableMap.of("var", "nam"),
+            i -> {
+                i.startRecord("1");
+                i.literal("name", "Max Power");
+                i.endRecord();
+
+                i.startRecord("2");
+                i.literal("name", "Some University");
+                i.endRecord();
+
+                i.startRecord("3");
+                i.literal("name", "Filibandrina");
+                i.endRecord();
+            },
+            o -> {
+                o.get().startRecord("1");
+                o.get().literal("name", "Max Power");
+                o.get().literal("type", "Person");
+                o.get().endRecord();
+
+                o.get().startRecord("2");
+                o.get().literal("name", "Some University");
+                o.get().literal("type", "Organization");
+                o.get().endRecord();
+
+                o.get().startRecord("3");
+                o.get().literal("name", "Filibandrina");
+                o.get().literal("type", "Unknown");
+                o.get().endRecord();
+            }
+        );
+    }
+
+    @Test
+    public void shouldResolveVariablesInUnless() {
+        MetafixTestHelpers.assertFix(streamReceiver, Arrays.asList(
+                "unless any_match('name', '.*Uni$[var]sity.*')",
+                "  add_field('type', 'Person')",
+                "end"
+            ),
+            ImmutableMap.of("var", "ver"),
+            i -> {
+                i.startRecord("1");
+                i.literal("name", "Max");
+                i.endRecord();
+
+                i.startRecord("2");
+                i.literal("name", "Some University");
+                i.endRecord();
+            },
+            o -> {
+                o.get().startRecord("1");
+                o.get().literal("name", "Max");
+                o.get().literal("type", "Person");
+                o.get().endRecord();
+
+                o.get().startRecord("2");
+                o.get().literal("name", "Some University");
+                o.get().endRecord();
+            }
+        );
+    }
+
 }

--- a/metafix/src/test/java/org/metafacture/metafix/MetafixScriptTest.java
+++ b/metafix/src/test/java/org/metafacture/metafix/MetafixScriptTest.java
@@ -23,7 +23,6 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
 import java.io.FileNotFoundException;
-import java.util.HashMap;
 import java.util.Map;
 import java.util.function.Consumer;
 
@@ -147,7 +146,7 @@ public class MetafixScriptTest {
 
     private void assertFix(final String fixDef, final Map<String, String> vars, final Consumer<Metafix> consumer) {
         try {
-            final Metafix metafix = vars == null ? new Metafix(fixDef) : new Metafix(fixDef, new HashMap<>(vars));
+            final Metafix metafix = vars == null ? new Metafix(fixDef) : new Metafix(fixDef, vars);
 
             // Prepare and trigger script execution
             metafix.setReceiver(new DefaultStreamReceiver());

--- a/metafix/src/test/java/org/metafacture/metafix/MetafixScriptTest.java
+++ b/metafix/src/test/java/org/metafacture/metafix/MetafixScriptTest.java
@@ -18,10 +18,13 @@ package org.metafacture.metafix;
 
 import org.metafacture.framework.helpers.DefaultStreamReceiver;
 
+import com.google.common.collect.ImmutableMap;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
 import java.io.FileNotFoundException;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.Consumer;
 
 /**
@@ -35,6 +38,48 @@ public class MetafixScriptTest {
     private static final String TSV_MAP = "src/test/resources/org/metafacture/metafix/maps/test.tsv";
 
     public MetafixScriptTest() {
+    }
+
+    @Test
+    public void shouldPutSingleVariable() {
+        assertVar("put_var('varName', 'value')",
+                null,
+                ImmutableMap.of("varName", "value"));
+    }
+
+    @Test
+    public void shouldPutMultipleVariables() {
+        assertVar("put_var('varName', 'value')\nput_var('varName2', 'value2')",
+                null,
+                ImmutableMap.of("varName", "value", "varName2", "value2"));
+    }
+
+    @Test
+    public void shouldPutMultipleVariablesFromMap() {
+        assertVar("put_vars(varName: 'value', varName2: 'value2')",
+                null,
+                ImmutableMap.of("varName", "value", "varName2", "value2"));
+    }
+
+    @Test
+    public void shouldResolveVariablesInSingleVariable() {
+        assertVar("put_var('varName', 'value$[var]')",
+                ImmutableMap.of("var", "1"),
+                ImmutableMap.of("varName", "value1"));
+    }
+
+    @Test
+    public void shouldResolveVariablesInMultipleVariables() {
+        assertVar("put_var('varName', 'value$[var]')\nput_var('$[varName]Var', 'value2')",
+                ImmutableMap.of("var", "1"),
+                ImmutableMap.of("varName", "value1", "value1Var", "value2"));
+    }
+
+    @Test
+    public void shouldNotResolveVariablesInMultipleVariablesFromMap() {
+        assertVar("put_vars(varName: 'value$[var]', '$[varName]Var': 'value2')",
+                ImmutableMap.of("var", "1"),
+                ImmutableMap.of("varName", "value$[var]", "$[varName]Var", "value2"));
     }
 
     @Test
@@ -83,6 +128,10 @@ public class MetafixScriptTest {
         assertMap("put_filemap('" + TSV_MAP + "', '" + MAP_NAME + "', sep_char: '\t')", MAP_NAME);
     }
 
+    private void assertVar(final String fixDef, final Map<String, String> vars, final Map<String, String> result) {
+        assertFix(fixDef, vars, f -> result.forEach((k, v) -> Assertions.assertEquals(v, f.getVars().get(k))));
+    }
+
     private void assertMap(final String fixDef, final String mapName) {
         assertFix(fixDef, f -> assertMap(f, mapName));
     }
@@ -93,8 +142,12 @@ public class MetafixScriptTest {
     }
 
     private void assertFix(final String fixDef, final Consumer<Metafix> consumer) {
+        assertFix(fixDef, null, consumer);
+    }
+
+    private void assertFix(final String fixDef, final Map<String, String> vars, final Consumer<Metafix> consumer) {
         try {
-            final Metafix metafix = new Metafix(fixDef);
+            final Metafix metafix = vars == null ? new Metafix(fixDef) : new Metafix(fixDef, new HashMap<>(vars));
 
             // Prepare and trigger script execution
             metafix.setReceiver(new DefaultStreamReceiver());

--- a/metafix/src/test/java/org/metafacture/metafix/MetafixTest.java
+++ b/metafix/src/test/java/org/metafacture/metafix/MetafixTest.java
@@ -39,6 +39,14 @@ public class MetafixTest {
     }
 
     @Test
+    public void shouldPutVar() {
+        final Metafix metafix = new Metafix();
+        metafix.putVar(KEY, VALUE);
+
+        Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
+    }
+
+    @Test
     public void shouldGetMapNames() {
         final Metafix metafix = new Metafix();
         metafix.putMap(MAP_NAME, new HashMap<>());

--- a/metafix/src/test/java/org/metafacture/metafix/MetafixTest.java
+++ b/metafix/src/test/java/org/metafacture/metafix/MetafixTest.java
@@ -21,8 +21,10 @@ import org.metafacture.metamorph.api.Maps;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
+import java.io.FileNotFoundException;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -44,6 +46,40 @@ public class MetafixTest {
         metafix.putVar(KEY, VALUE);
 
         Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
+    }
+
+    @Test
+    public void shouldPutVarWithMutableMap() {
+        final Map<String, String> map = new HashMap<>();
+        map.put(KEY, VALUE);
+
+        try {
+            final Metafix metafix = new Metafix("vacuum()", map);
+            metafix.putVar(KEY + "2", VALUE + "2");
+
+            Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
+            Assertions.assertEquals(VALUE + "2", metafix.getVars().get(KEY + "2"));
+        }
+        catch (final FileNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void shouldPutVarWithImmutableMap() {
+        final Map<String, String> map = new HashMap<>();
+        map.put(KEY, VALUE);
+
+        try {
+            final Metafix metafix = new Metafix("vacuum()", Collections.unmodifiableMap(map));
+            metafix.putVar(KEY + "2", VALUE + "2");
+
+            Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
+            Assertions.assertEquals(VALUE + "2", metafix.getVars().get(KEY + "2"));
+        }
+        catch (final FileNotFoundException e) {
+            throw new RuntimeException(e);
+        }
     }
 
     @Test

--- a/metafix/src/test/java/org/metafacture/metafix/MetafixTest.java
+++ b/metafix/src/test/java/org/metafacture/metafix/MetafixTest.java
@@ -21,7 +21,6 @@ import org.metafacture.metamorph.api.Maps;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
-import java.io.FileNotFoundException;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
@@ -43,7 +42,7 @@ public class MetafixTest {
     @Test
     public void shouldPutVar() {
         final Metafix metafix = new Metafix();
-        metafix.putVar(KEY, VALUE);
+        metafix.getVars().put(KEY, VALUE);
 
         Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
     }
@@ -53,16 +52,11 @@ public class MetafixTest {
         final Map<String, String> map = new HashMap<>();
         map.put(KEY, VALUE);
 
-        try {
-            final Metafix metafix = new Metafix("vacuum()", map);
-            metafix.putVar(KEY + "2", VALUE + "2");
+        final Metafix metafix = new Metafix(map);
+        metafix.getVars().put(KEY + "2", VALUE + "2");
 
-            Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
-            Assertions.assertEquals(VALUE + "2", metafix.getVars().get(KEY + "2"));
-        }
-        catch (final FileNotFoundException e) {
-            throw new RuntimeException(e);
-        }
+        Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
+        Assertions.assertEquals(VALUE + "2", metafix.getVars().get(KEY + "2"));
     }
 
     @Test
@@ -70,16 +64,11 @@ public class MetafixTest {
         final Map<String, String> map = new HashMap<>();
         map.put(KEY, VALUE);
 
-        try {
-            final Metafix metafix = new Metafix("vacuum()", Collections.unmodifiableMap(map));
-            metafix.putVar(KEY + "2", VALUE + "2");
+        final Metafix metafix = new Metafix(Collections.unmodifiableMap(map));
+        metafix.getVars().put(KEY + "2", VALUE + "2");
 
-            Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
-            Assertions.assertEquals(VALUE + "2", metafix.getVars().get(KEY + "2"));
-        }
-        catch (final FileNotFoundException e) {
-            throw new RuntimeException(e);
-        }
+        Assertions.assertEquals(VALUE, metafix.getVars().get(KEY));
+        Assertions.assertEquals(VALUE + "2", metafix.getVars().get(KEY + "2"));
     }
 
     @Test

